connectors/bigquery/flink-connector-gcp-bigquery/src/main/java/com/google/flink/connector/gcp/bigquery/BigQueryCatalog.java

```diff
@@ -0,0 +1,440 @@
+                    .setApplicationName("BigQuery Flink Catalog Plugin")
+                    .build();
+        } catch (GeneralSecurityException | IOException e) {
+            throw new CatalogException("Failed to create BigQuery client", e);
+        }
+    }
+
+    @Override
+    public void close() throws CatalogException {
+        if (client != null && client.getRequestFactory().getTransport() instanceof com.google.api.client.http.javanet.NetHttpTransport) {
+            try {
+                this.client.getRequestFactory().getTransport().shutdown();
+                LOG.info("Shutdown the underlying HTTP transport for BigQuery client.");
+            } catch (IOException e) {
+                throw new CatalogException("Failed to shutdown the HTTP transport", e);
+            } finally {
+                client = null;
+            }
+        } else if (client != null) {
+            client = null;
+            LOG.info("BigQuery client was open, but using a transport without explicit shutdown. Nullifying client.");
+        } else {
+            LOG.info("BigQuery client was not open, nothing to close.");
+        }
+    }
+
+    // -- Database Operations --
+    @Override
+    public List<String> listDatabases() throws CatalogException {
+        List<String> targetReturnList = new ArrayList<>();
+        try {
+            Datasets.List listDatasets = this.client.datasets().list(this.projectId);
+            DatasetList datasets = listDatasets.execute();
+
+            if (datasets == null || datasets.getDatasets() == null) {
+                LOG.debug("Project does not contain any datasets.");
+                return List.of();
+            }
+            datasets.getDatasets().forEach(dataset
+                    -> targetReturnList.add(dataset.getDatasetReference().getDatasetId()));
+            return targetReturnList;
+        } catch (IOException e) {
+            throw new CatalogException("Failed to list databases", e);
+        }
+
+    }
+
+    @Override
+    public CatalogDatabase getDatabase(String databaseName) throws DatabaseNotExistException, CatalogException {
+        try {
+            Datasets.Get getDataset = this.client.datasets().get(this.projectId, databaseName);
+            Dataset dataset = getDataset.execute();
+            if (dataset == null) {
+                throw new DatabaseNotExistException(getName(), databaseName);
+            }
+
+            return new BigQueryCatalogDatabase(this.projectId, dataset);
+        } catch (IOException e) {
+            throw new CatalogException("Failed to get database " + databaseName, e);
+        }
+    }
+
+    @Override
+    public boolean databaseExists(String databaseName) throws CatalogException {
+        try {
+            Datasets.Get getDataset = this.client.datasets().get(this.projectId, databaseName);
+            Dataset dataset = getDataset.execute();
+            return dataset != null;
+        } catch (GoogleJsonResponseException e) {
```

Comment:
you can combine these two into a "catch or" instead of having two separate catches